File: app/src/main/java/activity/amigosecreto/db/ParticipanteDAO.java
package activity.amigosecreto.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import java.util.ArrayList;
import java.util.List;

public class ParticipanteDAO {
    private MySQLiteOpenHelper helper;
    private SQLiteDatabase database;

    public ParticipanteDAO(Context ctx) {
        helper = new MySQLiteOpenHelper(ctx);
    }

    public void open() throws SQLException {
        database = helper.getWritableDatabase();
    }

    public void close() {
        helper.close();
    }

    public void inserir(Participante p, int grupoId) {
        ContentValues values = new ContentValues();
        values.put(MySQLiteOpenHelper.COLUMN_NOME, p.getNome());
        values.put(MySQLiteOpenHelper.COLUMN_EMAIL, p.getEmail());
        values.put(MySQLiteOpenHelper.COLUMN_TELEFONE, p.getTelefone());
        values.put(MySQLiteOpenHelper.COLUMN_ENVIADO, 0);
        values.put(MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID, grupoId);
        long id = database.insert(MySQLiteOpenHelper.TABLE_PARTICIPANTE, null, values);
        p.setId((int) id);
    }

    public void remover(int id) {
        String idStr = String.valueOf(id);
        database.delete(MySQLiteOpenHelper.TABLE_PARTICIPANTE, MySQLiteOpenHelper.COLUMN_ID + " = ?", new String[]{idStr});
        database.delete(MySQLiteOpenHelper.TABLE_EXCLUSAO, MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID + " = ? OR " + MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID + " = ?", new String[]{idStr, idStr});
    }

    public void deletarTodosDoGrupo(int grupoId) {
        // Obter todos os IDs dos participantes do grupo para remover as exclusões também
        Cursor cursor = database.query(MySQLiteOpenHelper.TABLE_PARTICIPANTE, new String[]{MySQLiteOpenHelper.COLUMN_ID},
                MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID + " = ?", new String[]{String.valueOf(grupoId)}, null, null, null);
        if (cursor.moveToFirst()) {
            do {
                int id = cursor.getInt(0);
                String idStr = String.valueOf(id);
                database.delete(MySQLiteOpenHelper.TABLE_EXCLUSAO, MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID + " = ? OR " + MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID + " = ?", new String[]{idStr, idStr});
            } while (cursor.moveToNext());
        }
        cursor.close();
        database.delete(MySQLiteOpenHelper.TABLE_PARTICIPANTE, MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID + " = ?", new String[]{String.valueOf(grupoId)});
    }

    public void limparSorteioDoGrupo(int grupoId) {
        ContentValues values = new ContentValues();
        values.putNull(MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID);
        values.put(MySQLiteOpenHelper.COLUMN_ENVIADO, 0);
        database.update(MySQLiteOpenHelper.TABLE_PARTICIPANTE, values, MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID + " = ?", new String[]{String.valueOf(grupoId)});
    }

    public void adicionarExclusao(int idParticipante, int idExcluido) {
        ContentValues values = new ContentValues();
        values.put(MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID, idParticipante);
        values.put(MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID, idExcluido);
        database.insertWithOnConflict(MySQLiteOpenHelper.TABLE_EXCLUSAO, null, values, SQLiteDatabase.CONFLICT_IGNORE);
    }

    public void removerExclusao(int idParticipante, int idExcluido) {
        database.delete(MySQLiteOpenHelper.TABLE_EXCLUSAO, 
                MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID + " = ? AND " + MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID + " = ?",
                new String[]{String.valueOf(idParticipante), String.valueOf(idExcluido)});
    }

    public boolean salvarSorteio(List<Participante> participantes, List<Participante> sorteados) {
        database.beginTransaction();
        try {
            for (int i = 0; i < participantes.size(); i++) {
                ContentValues values = new ContentValues();
                values.put(MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID, sorteados.get(i).getId());
                values.put(MySQLiteOpenHelper.COLUMN_ENVIADO, 0);
                database.update(MySQLiteOpenHelper.TABLE_PARTICIPANTE, values,
                        MySQLiteOpenHelper.COLUMN_ID + " = ?", new String[]{String.valueOf(participantes.get(i).getId())});
            }
            database.setTransactionSuccessful();
            return true;
        } catch (Exception e) {
            return false;
        } finally {
            database.endTransaction();
        }
    }

    public void marcarComoEnviado(int id) {
        ContentValues values = new ContentValues();
        values.put(MySQLiteOpenHelper.COLUMN_ENVIADO, 1);
        database.update(MySQLiteOpenHelper.TABLE_PARTICIPANTE, values, MySQLiteOpenHelper.COLUMN_ID + " = ?", new String[]{String.valueOf(id)});
    }

    public List<Participante> listarPorGrupo(int grupoId) {
        List<Participante> lista = new ArrayList<>();
        Cursor cursor = database.query(MySQLiteOpenHelper.TABLE_PARTICIPANTE, null, 
                MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID + " = ?", new String[]{String.valueOf(grupoId)}, null, null, MySQLiteOpenHelper.COLUMN_NOME);
        
        if (cursor.moveToFirst()) {
            do {
                Participante p = new Participante();
                p.setId(cursor.getInt(0));
                p.setNome(cursor.getString(1));
                p.setEmail(cursor.getString(2));
                p.setTelefone(cursor.getString(3));
                if (!cursor.isNull(4)) {
                    p.setAmigoSorteadoId(cursor.getInt(4));
                }
                p.setEnviado(cursor.getInt(5) == 1);
                
                // Carregar exclusões
                p.setIdsExcluidos(listarExclusoes(p.getId()));
                
                lista.add(p);
            } while (cursor.moveToNext());
        }
        cursor.close();
        return lista;
    }

    private List<Integer> listarExclusoes(int idParticipante) {
        List<Integer> exclusoes = new ArrayList<>();
        Cursor cursor = database.query(MySQLiteOpenHelper.TABLE_EXCLUSAO, 
                new String[]{MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID}, 
                MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID + " = ?", 
                new String[]{String.valueOf(idParticipante)}, null, null, null);
        if (cursor.moveToFirst()) {
            do {
                exclusoes.add(cursor.getInt(0));
            } while (cursor.moveToNext());
        }
        cursor.close();
        return exclusoes;
    }

    public String getNomeAmigoSorteado(int amigoId) {
        String nome = "Ninguém";
        Cursor cursor = database.query(MySQLiteOpenHelper.TABLE_PARTICIPANTE,
                new String[]{MySQLiteOpenHelper.COLUMN_NOME},
                MySQLiteOpenHelper.COLUMN_ID + " = ?",
                new String[]{String.valueOf(amigoId)},
                null, null, null);

        if (cursor.moveToFirst()) {
            nome = cursor.getString(0);
        }
        cursor.close();
        return nome;
    }

    public Participante buscarPorId(int id) {
        Participante p = null;
        Cursor cursor = database.query(MySQLiteOpenHelper.TABLE_PARTICIPANTE, null,
                MySQLiteOpenHelper.COLUMN_ID + " = ?",
                new String[]{String.valueOf(id)},
                null, null, null);

        if (cursor.moveToFirst()) {
            p = new Participante();
            p.setId(cursor.getInt(0));
            p.setNome(cursor.getString(1));
            p.setEmail(cursor.getString(2));
            p.setTelefone(cursor.getString(3));
            if (!cursor.isNull(4)) {
                p.setAmigoSorteadoId(cursor.getInt(4));
            }
            p.setEnviado(cursor.getInt(5) == 1);
        }
        cursor.close();
        return p;
    }
}

File: app/src/main/java/activity/amigosecreto/db/ParticipanteDAO.kt
package activity.amigosecreto.db

import android.content.ContentValues
import android.content.Context
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.util.Log

class ParticipanteDAO(ctx: Context) {

    companion object {
        // TODO: move to strings.xml when RevelarAmigoActivity migrates to Kotlin (Fase 10e)
        //  and getNomeAmigoSorteado() can return String? instead.
        const val NOME_AMIGO_DESCONHECIDO = "Ninguém"
    }

    private val helper = MySQLiteOpenHelper(ctx)
    private lateinit var database: SQLiteDatabase

    @Throws(SQLException::class)
    fun open() {
        database = helper.writableDatabase
    }

    fun close() {
        helper.close()
    }

    fun inserir(p: Participante, grupoId: Int) {
        val values = ContentValues().apply {
            put(MySQLiteOpenHelper.COLUMN_NOME, p.nome)
            put(MySQLiteOpenHelper.COLUMN_EMAIL, p.email)
            put(MySQLiteOpenHelper.COLUMN_TELEFONE, p.telefone)
            put(MySQLiteOpenHelper.COLUMN_ENVIADO, 0)
            put(MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID, grupoId)
        }
        val id = database.insert(MySQLiteOpenHelper.TABLE_PARTICIPANTE, null, values)
        p.id = id.toInt() // safe: participant count never approaches Int.MAX_VALUE (2^31-1)
    }

    fun atualizar(p: Participante): Boolean {
        if (p.id <= 0) return false
        val values = ContentValues().apply {
            put(MySQLiteOpenHelper.COLUMN_NOME, p.nome)
            put(MySQLiteOpenHelper.COLUMN_EMAIL, p.email)
            put(MySQLiteOpenHelper.COLUMN_TELEFONE, p.telefone)
            // amigo_sorteado_id e enviado são preservados intencionalmente
        }
        val rows = database.update(
            MySQLiteOpenHelper.TABLE_PARTICIPANTE, values,
            "${MySQLiteOpenHelper.COLUMN_ID} = ?", arrayOf(p.id.toString())
        )
        return rows > 0
    }

    fun remover(id: Int) {
        // ON DELETE CASCADE (schema v9) handles exclusao and desejo deletion automatically.
        database.delete(MySQLiteOpenHelper.TABLE_PARTICIPANTE, "${MySQLiteOpenHelper.COLUMN_ID} = ?", arrayOf(id.toString()))
    }

    fun deletarTodosDoGrupo(grupoId: Int) {
        // ON DELETE CASCADE (schema v9) handles exclusao and desejo deletion automatically.
        database.delete(MySQLiteOpenHelper.TABLE_PARTICIPANTE, "${MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID} = ?", arrayOf(grupoId.toString()))
    }

    fun atualizarAmigoSorteado(participanteId: Int, amigoSorteadoId: Int) {
        val values = ContentValues().apply {
            put(MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID, amigoSorteadoId)
        }
        database.update(MySQLiteOpenHelper.TABLE_PARTICIPANTE, values,
            "${MySQLiteOpenHelper.COLUMN_ID} = ?", arrayOf(participanteId.toString()))
    }

    fun limparSorteioDoGrupo(grupoId: Int) {
        val values = ContentValues().apply {
            putNull(MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID)
            put(MySQLiteOpenHelper.COLUMN_ENVIADO, 0)
        }
        database.update(MySQLiteOpenHelper.TABLE_PARTICIPANTE, values, "${MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID} = ?", arrayOf(grupoId.toString()))
    }

    fun adicionarExclusao(idParticipante: Int, idExcluido: Int) {
        val values = ContentValues().apply {
            put(MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID, idParticipante)
            put(MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID, idExcluido)
        }
        database.insertWithOnConflict(MySQLiteOpenHelper.TABLE_EXCLUSAO, null, values, SQLiteDatabase.CONFLICT_IGNORE)
    }

    fun removerExclusao(idParticipante: Int, idExcluido: Int) {
        database.delete(
            MySQLiteOpenHelper.TABLE_EXCLUSAO,
            "${MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID} = ? AND ${MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID} = ?",
            arrayOf(idParticipante.toString(), idExcluido.toString())
        )
    }

    /**
     * Aplica todas as alterações de exclusão de um participante em uma única transação atômica.
     * Evita falha parcial e múltiplos open/close que ocorreriam num loop de chamadas individuais.
     */
    fun salvarExclusoes(participanteId: Int, adicionar: List<Int>, remover: List<Int>) {
        database.beginTransaction()
        try {
            for (id in adicionar) {
                val values = ContentValues().apply {
                    put(MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID, participanteId)
                    put(MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID, id)
                }
                database.insertWithOnConflict(MySQLiteOpenHelper.TABLE_EXCLUSAO, null, values, SQLiteDatabase.CONFLICT_IGNORE)
            }
            for (id in remover) {
                database.delete(
                    MySQLiteOpenHelper.TABLE_EXCLUSAO,
                    "${MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID} = ? AND ${MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID} = ?",
                    arrayOf(participanteId.toString(), id.toString())
                )
            }
            database.setTransactionSuccessful()
        } finally {
            database.endTransaction()
        }
    }

    fun salvarSorteio(participantes: List<Participante>, sorteados: List<Participante>): Boolean {
        if (participantes.size != sorteados.size) {
            Log.e("ParticipanteDAO", "salvarSorteio: list size mismatch (${participantes.size} vs ${sorteados.size})")
            return false
        }
        database.beginTransaction()
        try {
            for (i in participantes.indices) {
                val values = ContentValues().apply {
                    put(MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID, sorteados[i].id)
                    put(MySQLiteOpenHelper.COLUMN_ENVIADO, 0)
                }
                database.update(
                    MySQLiteOpenHelper.TABLE_PARTICIPANTE, values,
                    "${MySQLiteOpenHelper.COLUMN_ID} = ?", arrayOf(participantes[i].id.toString())
                )
            }
            database.setTransactionSuccessful()
            return true
        } catch (e: SQLiteException) {
            Log.e("ParticipanteDAO", "salvarSorteio failed", e)
            return false
        } finally {
            database.endTransaction()
        }
    }

    fun marcarComoEnviado(id: Int) {
        val values = ContentValues().apply {
            put(MySQLiteOpenHelper.COLUMN_ENVIADO, 1)
        }
        database.update(MySQLiteOpenHelper.TABLE_PARTICIPANTE, values, "${MySQLiteOpenHelper.COLUMN_ID} = ?", arrayOf(id.toString()))
    }

    fun listarPorGrupo(grupoId: Int): List<Participante> {
        // Pass 1 — fetch participants
        val mapaParticipantes = linkedMapOf<Int, Participante>()
        val cursor = database.query(
            MySQLiteOpenHelper.TABLE_PARTICIPANTE, null,
            "${MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID} = ?", arrayOf(grupoId.toString()),
            null, null, MySQLiteOpenHelper.COLUMN_NOME
        )
        cursor.use {
            if (it.moveToFirst()) {
                val idIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_ID)
                val nomeIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_NOME)
                val emailIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_EMAIL)
                val telefoneIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_TELEFONE)
                val amigoIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID)
                val enviadoIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_ENVIADO)
                do {
                    val p = Participante()
                    p.id = it.getInt(idIdx)
                    p.nome = it.getString(nomeIdx)
                    p.email = it.getString(emailIdx)
                    p.telefone = it.getString(telefoneIdx)
                    if (!it.isNull(amigoIdx)) p.amigoSorteadoId = it.getInt(amigoIdx)
                    p.isEnviado = it.getInt(enviadoIdx) == 1
                    mapaParticipantes[p.id] = p
                } while (it.moveToNext())
            }
        }
        if (mapaParticipantes.isEmpty()) return emptyList()

        // Pass 2 — fetch all exclusions for the group in one query (eliminates N+1).
        // IDs are integers from our own DB — no SQL injection risk. rawQuery with '?'
        // does not support dynamic IN lists, so interpolation is the correct approach here.
        val ids = mapaParticipantes.keys.joinToString(",")
        val exclCursor = database.rawQuery(
            "SELECT ${MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID}, ${MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID}" +
            " FROM ${MySQLiteOpenHelper.TABLE_EXCLUSAO}" +
            " WHERE ${MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID} IN ($ids)",
            null
        )
        exclCursor.use {
            if (it.moveToFirst()) {
                val pidIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_PARTICIPANTE_ID)
                val eidIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_EXCLUIDO_ID)
                do {
                    mapaParticipantes[it.getInt(pidIdx)]?.idsExcluidos?.add(it.getInt(eidIdx))
                } while (it.moveToNext())
            }
        }

        return mapaParticipantes.values.toList()
    }

    fun getNomeAmigoSorteado(amigoId: Int): String {
        val cursor = database.query(
            MySQLiteOpenHelper.TABLE_PARTICIPANTE,
            arrayOf(MySQLiteOpenHelper.COLUMN_NOME),
            "${MySQLiteOpenHelper.COLUMN_ID} = ?",
            arrayOf(amigoId.toString()),
            null, null, null
        )
        // "Ninguém" is a DAO-level fallback; ideally the ViewModel would own this string
        // (via getString R.string.*), but changing the return type cascades to RevelarAmigoActivity.java
        // which calls the DAO directly. TODO: migrate to String? return in Fase 10e when
        // RevelarAmigoActivity migrates to Kotlin.
        return cursor.use {
            if (it.moveToFirst()) it.getString(0) else NOME_AMIGO_DESCONHECIDO
        }
    }

    fun contarPorGrupo(): Map<Int, Int> {
        val mapa = mutableMapOf<Int, Int>()
        val cursor = database.rawQuery(
            "SELECT ${MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID}, COUNT(*) AS cnt" +
            " FROM ${MySQLiteOpenHelper.TABLE_PARTICIPANTE}" +
            " GROUP BY ${MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID}",
            null
        )
        cursor.use {
            if (it.moveToFirst()) {
                val grupoIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID)
                val cntIdx = it.getColumnIndexOrThrow("cnt")
                do { mapa[it.getInt(grupoIdx)] = it.getInt(cntIdx) } while (it.moveToNext())
            }
        }
        return mapa
    }

    /**
     * Retorna o número de grupos que têm >= [minParticipantes] participantes mas onde
     * nenhum participante possui amigo_sorteado_id definido (sorteio ainda não realizado).
     *
     * Usa GROUP BY + HAVING numa única query, evitando N+1.
     * Padrão equivalente a [contarPorGrupo] e ao batch de DesejoDAO.
     *
     * Nota: [minParticipantes] é um Int interno injetado como literal no SQL (não como ?).
     * rawQuery com ? dentro de subquery tem comportamento inconsistente no Robolectric 4.13;
     * o valor não vem de entrada externa, portanto não há risco de SQL injection.
     */
    fun contarGruposPendentes(minParticipantes: Int = 3): Int {
        // COUNT(coluna) conta apenas valores não-NULL — se todos amigo_sorteado_id são NULL,
        // COUNT = 0, indicando que o sorteio ainda não foi realizado.
        val sql =
            "SELECT ${MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID}" +
            " FROM ${MySQLiteOpenHelper.TABLE_PARTICIPANTE}" +
            " GROUP BY ${MySQLiteOpenHelper.COLUMN_FK_GRUPO_ID}" +
            " HAVING COUNT(*) >= $minParticipantes" +
            "   AND COUNT(${MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID}) = 0"
        val cursor = database.rawQuery(sql, null)
        return cursor.use { it.count }
    }

    /**
     * Retorna o participante pelo id, ou null se não existir.
     * Nota: [Participante.idsExcluidos] NÃO é populado — use [listarPorGrupo] quando
     * as exclusões forem necessárias.
     */
    fun buscarPorId(id: Int): Participante? {
        val cursor = database.query(
            MySQLiteOpenHelper.TABLE_PARTICIPANTE, null,
            "${MySQLiteOpenHelper.COLUMN_ID} = ?", arrayOf(id.toString()),
            null, null, null
        )
        return cursor.use {
            if (!it.moveToFirst()) return null
            val p = Participante()
            p.id = it.getInt(it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_ID))
            p.nome = it.getString(it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_NOME))
            p.email = it.getString(it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_EMAIL))
            p.telefone = it.getString(it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_TELEFONE))
            val amigoIdx = it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_AMIGO_SORTEADO_ID)
            if (!it.isNull(amigoIdx)) p.amigoSorteadoId = it.getInt(amigoIdx)
            p.isEnviado = it.getInt(it.getColumnIndexOrThrow(MySQLiteOpenHelper.COLUMN_ENVIADO)) == 1
            p
        }
    }
}
